### PR TITLE
Check select

### PR DESCRIPTION
**Pull request by Devansh-Patel**
This PR fixes the issue of select system call not checking whether the pipe is ready to read/write or not.

### CODE REVIEW

File: src/interface/pipe.rs [hunk 1]
```diff
@@ -58,6 +58,29 @@ impl EmulatedPipe {
         if (flags & O_RDWRFLAGS) == O_RDONLY {self.refcount_read.fetch_sub(1, Ordering::Relaxed);}
         if (flags & O_RDWRFLAGS) == O_WRONLY {self.refcount_write.fetch_sub(1, Ordering::Relaxed);}
     }
+    pub fn check_select_read(&self) -> bool {
+        let read_end = self.read_end.lock();
+        let pipe_space = read_end.len();
+
+        if (pipe_space > 0) && !self.eof.load(Ordering::SeqCst){
```
Comment (jesings) [inline src/interface/pipe.rs:65]: Why do we check EOF here?
How can there be stuff in the read end of the pipe if we've already reached an EOF.
If the EOF is after the stuff in the read end, then select should still exit.

File: src/interface/pipe.rs [hunk 2]
```diff
@@ -58,6 +58,29 @@ impl EmulatedPipe {
...
+            return false;
+        }
+    }
+    pub fn check_select_write(&self) -> bool {
+
+        let write_end = self.write_end.lock();
+
+        let pipe_space = write_end.remaining();
+        if pipe_space == self.size {
+            return false;
```
Comment (jesings) [inline src/interface/pipe.rs:78]: Don't we want to check if the remaining space is nonzero?

**Review by jesings — CHANGES_REQUESTED** [review]
I don't think this is semantically correct, but the structure looks good.

File: src/interface/pipe.rs (hunk 1 above)
Comment (rennergade) [reply src/interface/pipe.rs:65]: I believe this actually should be an OR and just a eof load (non-negated).

One the EOF is set, select will always say the pipe is readable, even if theres nothing in it. But before its set, its only readable when there is data in the pipe.

Comment (Devansh-Patel) [reply src/interface/pipe.rs:65]: Noted.

File: src/interface/pipe.rs (hunk 2 above)
Comment (Devansh-Patel) [reply src/interface/pipe.rs:78]: Noted.

File: src/interface/pipe.rs [hunk 3]
```diff
@@ -58,6 +58,29 @@ impl EmulatedPipe {
...
+
+        let write_end = self.write_end.lock();
+
+        let pipe_space = write_end.remaining();
+        if pipe_space == 0 {
+            return false;
+        }
+        else {
+            return true;
+        }
```
Comment (jesings) [inline src/interface/pipe.rs:82]: This could just be replaced by `return pipe_space != 0`, which I think is cleaner, but this isn't a big deal.

**Review by jesings — APPROVED** [review]
One minor code cleanup thing, but overall this looks great!

File: src/interface/pipe.rs (hunk 3 above)
Comment (Devansh-Patel) [reply src/interface/pipe.rs:82]: Affirmative.

**Review by rennergade — APPROVED** [review]

**Pull request merged**